keps/sig-windows/20190103-windows-node-support.md

```diff
@@ -173,6 +179,19 @@ Windows nodes will adhere to Kubernetes version-skew policy (node to control pla
 - [cncf-k8s-conformance thread](https://lists.cncf.io/g/cncf-k8s-conformance/topic/windows_conformance_tests/27913232)
 - [kubernetes/enhancements proposal](https://github.com/kubernetes/features/issues/116)
 
+### API Changes Proposed
+
+#### Add V1.SecurityContext.WindowsOptions.runAsUsername
+
+Linux uses userID (UID) and groupID (GID) which are represented as integer types. User and group names are not canonical - they are just an alias in /etc/groups or /etc/passwd back to UID+GID. Windows uses a larger binary security identifier (SID) which is stored in the Windows Security Access Manager (SAM) database. This database is not shared between the host and containers, or between containers.
+
+The Windows OCI spec and CRI can handle a username instead of UID, which is interpreted inside the container to create a process as the intended user. [64009](https://github.com/kubernetes/kubernetes/pull/64009) added this to the CRI and initially proposed changing the existing `V1.Container.SecurityContext.runAsUser` field to IntStr in a later PR. More discussion on [#73387](https://github.com/kubernetes/kubernetes/issues/73387) showed that would not work, so we're proposing to add `SecurityContext.WindowsOptions.RunAsUserName` as an alternative to support Windows users as strings. As [GMSA](https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20181221-windows-group-managed-service-accounts-for-container-identity.md) graduates to beta, the field `CredentialSpec` could be added to the same object as well.
+
+On Windows, if `runAsUser` is set today, the container will not start with an error "run as uid (%d) is not supported on Windows" [src](https://github.com/kubernetes/kubernetes/blob/a5ade16abd07b938978d4876e4909b956d25c223/pkg/kubelet/kuberuntime/kuberuntime_container_windows.go#L88). This behavior will not change. Setting `runAsGroup` or `supplementalGroups` without `runAsUser` is invalid so the existing check is sufficient on Windows.
+
+The new field `runAsUserName` would not be implemented on Linux. Although the field `run_as_username` exists in CRI and the OCI container spec for Linux, it was [intended to be deprecated](https://github.com/kubernetes/kubernetes/pull/36542#discussion_r87922802) from CRI. It was never hooked up in the Kubernetes API and dockershim. We propose leaving it in the state it's in today and not adding this as a new feature on Linux. We can add a check at the same time so that if it is set on a Linux container, an error "run as username is not supported on Linux".
```

Comment:
iiuc, specifying a string username when starting a linux container doesn't actually give control over the user the container runs as, since it still goes through a layer of indirection via /etc/passwd under the container image's control.

Is there a reason not to stick with the minimal, canonical user specification for each platform (int uid for linux, string username for windows)?